vent/coordinator/ipc.py

```diff
@@ -1,28 +1,90 @@
 import threading
 import time
+from enum import Enum, auto
+import pickle
+import socket
+
+buffer_size = 65536
+
+
+class IPCCommand(Enum):
+    START = auto()
+    STOP = auto()
+    GET_SENSORS = auto()
+    GET_CONTROL = auto()
+    SET_CONTROL = auto()
+    GET_ACTIVE_ALARMS = auto()
+    GET_LOGGED_ALARMS = auto()
+    CLEAR_LOGGED_ALARMS = auto()
+
+
+class IPCMessage:
+    def __init__(self, command, args=None):
+        """
+        :param command: ENUM in IPCCommand
+        """
+        self.command = command
+        # TODO: current only assume one arg
+        self.args = args
 
 
 class IPC:
     # Class for communicating between processes either by sockets
     # or named-pipes
     #   Functions:
-    def __init__(self, listen=False):
+    def __init__(self, listen, addr=None, port=None):
+        self.listen = listen
+        self.s = socket.socket()
+        if listen:
+            # TODO: can start random port on testing, but deterministic port on deployment
+            self.s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            self.s.bind(('', 0))
+            self.addr, self.port = self.s.getsockname()
+            # print(self.addr)
+            # print(self.port)
+            # print('listen socket created')
+            self.conn = None
+            self.connection_thread = threading.Thread(target=self.connect, daemon=True)
+            self.connection_thread.start()
+        else:
+            self.s.connect((addr, port))
+            # print('connect socket created')
         # init as either listener or connector
-        self.thread = threading.Thread(target=self.start)
-        self.thread.start()
+        # self.thread = threading.Thread(target=self.start)
+        # self.thread.start()
+        # two msg queue
+
+    def connect(self):
+        self.s.listen()
+        self.conn, _ = self.s.accept()
 
     def __del__(self):
-        self.thread.join()
+        if self.listen:
+            self.conn.close()
+        self.s.close()
 
-    def sendMsg(self, msg):
-        pass
 
-    def recvMsg(self, timeout=None):
-        # set timeout=0 for non-blocking
-        pass
+    def send_msg(self, msg: IPCMessage):
+        # print(f'sending IPC: {msg.command}')
+        pickled_msg = pickle.dumps(msg)
+        if self.listen:
+            self.conn.sendall(pickled_msg)
```

Comment:
after `setblocking(True)`, all these commands need to be

```python
try:
    self.conn.sendall(pickled_msg)
except ExceptionName as e:
    handle_message_error(e)
```

Reply:
Changing to RPC solves this issue.